src/core/blocks/create_block.c

```diff
@@ -0,0 +1,165 @@
+  unsigned int num_included = 0;
+  MemPool *s;
+  unsigned int coinbase = 1;
+
+  unsigned int tx_desired = calc_num_tx_target();
+  unsigned int tx_in_mempool = HASH_COUNT(mempool);
+
+  //Take minimum value of the mempool or the desired size
+  unsigned int tx_to_get = ((tx_desired) < (tx_in_mempool)) ? (tx_desired) : (tx_in_mempool);
+
+  *tx_pts = malloc((tx_to_get+coinbase)*sizeof(Transaction**)); // Shouldn't this be multiplied by 4 because eeach pointer takes up 4 bytes?
+
+  for(s = mempool; s != NULL; s = s->hh.next) {
+    (*tx_pts)[num_included+1] = s->tx; // Leave the first tx for coinbase tx
+    tx_fees += calc_tx_fees(s->tx);
+    num_included++;
+    if(num_included == tx_to_get){
+      break;
+    }
+  }
+
+  // Create the Coinbase TX
+  (*tx_pts)[0] = create_coinbase_tx(tx_fees);
+  
+  unsigned int total_num_txs = num_included+coinbase;
+  // Resize the Transaction memory if the wrong size was created
+  if(num_included < tx_to_get){
+    void *tmp = realloc(*tx_pts, (total_num_txs)*sizeof(Transaction**));
+    if (tmp == NULL) {
+        //realloc didn't work  still points to the og location
+    }
+    else{
+      *tx_pts = tmp;
+    }
+  }
+  return total_num_txs;
+}
+
+unsigned char *hash_all_tx(Transaction **txs, unsigned int num_txs){
+  unsigned char *total_hash = malloc(ALL_TX_HASH_LEN);
+  unsigned char *temp_all = malloc(TX_HASH_LEN*num_txs);
+  
+  for(unsigned int i = 0; i < num_txs; i++){
+    hash_tx(temp_all+i*TX_HASH_LEN, (txs)[i]);
+  }
+
+  hash_sha256(total_hash, temp_all, TX_HASH_LEN*num_txs);
+  free(temp_all);
+  return total_hash;
+}
+
+BlockHeader *create_block_header(Transaction **txs, unsigned int num_txs){
+  BlockHeader *b_header = malloc(sizeof(BlockHeader));
+  unsigned char *prev_header_hash = get_prev_header_hash();
+  unsigned char *all_tx_hash = hash_all_tx(txs, num_txs);
+
+  memcpy(b_header->prev_header_hash, prev_header_hash, ALL_TX_HASH_LEN); 
+  memcpy(b_header->all_tx, all_tx_hash, ALL_TX_HASH_LEN); 
+  b_header->nonce = 0;
+  b_header->timestamp = time(NULL);
+  
+  return b_header;
+}
+
+void create_block(Block *block){
+  Transaction **tx_ptr = malloc(sizeof(Transaction**));
+  block->num_txs = get_txs_from_mempool(&tx_ptr);
+  block->header = *create_block_header(tx_ptr, block->num_txs);
+  block->txs = tx_ptr; 
+}
+
+Block *create_block_alloc(){
+  Block *block = malloc(sizeof(Block));
+  create_block(block);
+  return block;
+}
+
+void change_nonce(Block *block){
+  block->header.nonce += 1;
+}
+
+int try_header_hash(BlockHeader *block_header){
+  // Hash the header
+  unsigned char header_hash[BLOCK_HASH_LEN];
+  hash_blockheader(header_hash, block_header);
+  unsigned char validator[HASH_DIFFICULTY];
+  memset(validator, 0, HASH_DIFFICULTY);
+  int ret = memcmp(validator, header_hash, HASH_DIFFICULTY);
+  if(ret != 0){
+    return 1; // Failure, bad hash
+  }
+  else{
+    return 0; // Difficulty Proof of work met
+  }
+}
+
+int mine_block(Block *block){
```

Comment:
This should probably also create the block

Reply:
sorry, this function wasn't well thoguht out, it was litteraly just the same as a header hash, but passing a block for convenience. This is likely useless, I held off on writing a mining loop because that will lock up our process. I will rewrite this to a mining loop that you are mentioning